static/download-service.ts

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2023, Compiler Explorer Authors
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright notice,
+//       this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+import $ from 'jquery';
+
+import {FiledataPair} from '../types/compilation/compilation.interfaces';
+
+export type SourceAndFiles = {
+    files: FiledataPair[];
+    source: string;
+};
+
+export type FileToDownload = {
+    url: string;
+    filename: string;
+    contents?: string;
+};
+
+export class IncludeDownloads {
+    private toDownload: FileToDownload[] = [];
+    private downloadPromises: Promise<FileToDownload>[] = [];
+
+    private async startDownload(download) {
+        this.downloadPromises.push(
+            new Promise(resolve => {
+                $.get(download.url, data => {
+                    download.contents = data;
+                    resolve(download);
+                });
+            })
+        );
+    }
+
+    private getFilenameFromUrl(url) {
+        const jsurl = new URL(url);
+        const urlpath = jsurl.pathname;
+        return jsurl.host + urlpath;
```

Comment:
we may have to sanitise this slightly? Nothing should escape the jail but funky filenames like `http://some.org/../../../../../../something` might wind up in a weird spot. I think that would map to `some.org/../../../../..something`, right?

My instinct would be to give each downloaded file a unique name in a "known" directory. But we can stick to this perhaps and replace all non `a-z0-9_` with `_` or something? so my example above would be `some.org_____________something`

Reply:
Would .. be a valid url path segment? I feel like webservers might just return a 500, will need to test

Reply:
doesn't this mean the file will be written to disk as that path, and then the `#include` would be that? I'm not worried about the remote fetch, just the oddity of the filename being (possibly) invalid on the host file system. Or did I miss that something else happens to the filename? Definitely not worried about the security here, I trust the sandbox, but I can imagine things getting weird :)

i guess not worth worrying about too much. just hope nobody includes `#include <http://blah.org/../example.cpp>` ;)

Reply:
ah, I did not realize the browser itself would rewrite this and you can make a valid url out of this, so yes, you're right

Reply:
Oh, except, the URL class already has provisions for this.
```
var url = new URL("http://www.google.com/helo/../../test.txt")
console.log(url.pathname);
```
puts out `/test.txt`

Reply:
The .pathname trick looks nice!

Reply:
oh neat! I retract my statement then!